### PR TITLE
Update navigation bar

### PR DESCRIPTION
**Pull request by TangerineFloof**
I focused on a couple QoL improvements for the navigation bar:

**Cleaned up links.** I've simplified the names and icons for the navbar
* "Information" → "About"
* "Updates" → "News"
    * The original name "Updates" felt more like "changelog" — you'd go to that page for software update information
    * "News" felt like it better conveyed that the page wasn't about software, it was about events/changes/updates in the community
* "+Join/Update" → "Fursuit Makers"
    * I did this to clarify that you can't sign up for a new user account with getfursu.it, it's just for fursuit makers

I also reorganized the links. Following web design principles that expect left → right, I put the links that are relevant for all users on the left. Links that are less universal are on the right.

I also updated the icons to try and improve clarity/use more familiar iconography.

**Removed "Maker IDs" link.** There's a link to it from the about page. Having it in the navbar competed for user's attention. I also think future UI updates can better highlight Maker IDs within the app, so there wouldn't be a significant need for a dedicated page.

**Made active page more visible.** The label is bold to show you which page you're on

**Unread events badge.** If a new event has been created since you last visited the events page, it will display a badge on the icon. The badge goes away when you visit the page.

## Screenshots

**Before:**
image

**After:**
image
image

### CODE REVIEW

File: symfony/src/Repository/EventRepository.php [hunk 1]
```diff
@@ -51,4 +53,19 @@ public function getRecent(array $types = []): array
...
     }
+
+    public function getLatestEventTimestamp(): ?DateTimeImmutable
+    {
+        $resultData = $this
+            ->createQueryBuilder('e')
+            ->select('MAX(e.timestamp)')
+            ->getQuery()
+            ->enableResultCache(3600)
+            ->getSingleScalarResult();
```
Comment (veelkoov) [inline symfony/src/Repository/EventRepository.php:64]: Since `null` could be a expected allowed scalar result, Doctrine throws `NoResultException`, which along with `NonUniqueResultException` inherits from `UnexpectedResultException`. Please catch `UnexpectedResultException` and return `null` in that case.

Comment (TangerineFloof) [reply symfony/src/Repository/EventRepository.php:64]: Oh, thank you for the catch!!

Comment (TangerineFloof) [reply symfony/src/Repository/EventRepository.php]: I wanted to investigate this so I could familiarize myself with PHP/the codebase more and figure out how to spot this bug myself in the future.

I checked out the source code for **doctrine/orm** and I think it isn't consistent in throwing this error. [`getSingleScalarResult`](https://github.com/doctrine/orm/blob/9c560713925ac5859342e6ff370c4c997acf2fd4/src/AbstractQuery.php#L797) does say it throws `NoResultException` if there's no result. But it only throws the error [if `hydrationMode` isn't `HYDRATE_SINGLE_SCALAR`](https://github.com/doctrine/orm/blob/9c560713925ac5859342e6ff370c4c997acf2fd4/src/AbstractQuery.php#L775). `executeIgnoreQueryCache` (if cache disabled or there's a cache miss) calls [`setHydrationMode`](https://github.com/doctrine/orm/blob/9c560713925ac5859342e6ff370c4c997acf2fd4/src/AbstractQuery.php#L905), but if the query is cached and there's a cache hit, then we don't call `setHydrationMode`. As a result, even though the documentation says it will throw the error, it doesn't throw if the query isn't cached.

I did this research because when I tried to cause the error intentionally (to verify it), I couldn't — `getSingleScalarResult` returned `null` when I had no rows in `EVENTS`.

I was also surprised to learn that even though I'd used `enableResultCache(3600)`, the query wasn't "cached" — it only ever called `executeIgnoreQueryCache`. You would need to call `setCacheable(true)` — but also, it seems the query cache isn't enabled?

**REGARDLESS!** I can't tell if this is intended behavior (documentation is wrong) or not (documentation is right and this is a bug). It's also what you want, so I've added the `try` statement. But I wanted to share this result because it seemed interesting?

**Comment by veelkoov** [comment]
What I've done using Bootstrap was not a standard solution, and now that next item has been added, the second row appears on top on mobile (which is most of getfursu.it visitors last time I saw a report):
image

File: symfony/src/Repository/EventRepository.php (hunk 1 above)
Comment (veelkoov) [reply symfony/src/Repository/EventRepository.php]: Cache:
- I am slowly mowing towards cache only on the controller/service level, the DB-level cache caused a big impact on memory; I was using two-level cache, this does not make much sense now that I think about it.
- It could be that the dev environment is disabling the cache whatsoever (for Doctrine, this possibly enables cache on prod: https://github.com/veelkoov/fuzzrake/blob/3e5132c9a1995ecb01891bafffd6e58a48fcf8f4/symfony/config/packages/doctrine.yaml#L40 ; I am unable to quickly find the settings for the framework)

Comment (veelkoov) [reply symfony/src/Repository/EventRepository.php]: Doctrine Behavior: weird... I created a test here: https://github.com/veelkoov/fuzzrake/commit/a1677f3aee6fbfa5a6c61ad802dcb8dacc988a71 (branch [testing/doctrine-behavior-test](https://github.com/veelkoov/fuzzrake/compare/testing/doctrine-behavior-test) and you can verify the results by `./toolbox pu --filter testGetSingleScalarResult`. Could you try running the test in your environment? I've never used Doctrine outside Symfony.

File: symfony/assets/scripts/class/Visitor.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    Storage.saveString(
+      STORAGE_KEY_LAST_VISITED_EVENTS_PAGE,
+      date.toISOString(),
+    );
+  }
+}
+
+export default Visitor;
```
Comment (veelkoov) [inline symfony/assets/scripts/class/Visitor.ts:37]: Question from a non-experienced dev. This is a single-purpose file. Why didn't you merge this into `export default class Visitor {`? Is this a convention?

File: symfony/assets/scripts/entry/general.ts [hunk 3]
```diff
@@ -32,3 +33,43 @@ jQuery(() => {
     config.save();
   });
 });
+
+jQuery(() => {
+  const newsNavlink = jQuery("#navlink-news");
```
Comment (veelkoov) [inline symfony/assets/scripts/entry/general.ts:38]: I use a convention of prepending any jQuery objects with `$`. Do you agree with this, or would you rather not? NOTE: this is not a convention coming from PHP variables.

File: symfony/assets/scripts/entry/general.ts [hunk 4]
```diff
@@ -32,3 +33,43 @@ jQuery(() => {
...
+  // the badge
+  if (lastVisitedEventsPage.valueOf() >= timestamp.valueOf()) {
+    return;
+  }
+
+  // There are new events. Show an unread badge
+  const icon = newsNavlink.find("i");
+  const badge = jQuery(document.createElement("div"));
+  badge.addClass("unread-badge");
+  icon.append(badge);
```
Comment (veelkoov) [inline symfony/assets/scripts/entry/general.ts:74]: Why wouldn't you use something like (writing from memory, may be broken code) `icon.append('<div class="unread-badge"/>');`? I suspect that your code could be faster, however this is done once per page load, so simplicity could be better here?

I also wonder, as a person not experienced on frontents, why is it OK to inject a block element into a line element, why wouldn't you use span instead of a div?

File: symfony/assets/scripts/entry/general.ts [hunk 5]
```diff
@@ -32,3 +33,43 @@ jQuery(() => {
...
+jQuery(() => {
+  const newsNavlink = jQuery("#navlink-news");
+
+  // If we're ON the events page, then never show a badge
+  if (newsNavlink.hasClass("active")) {
+    return;
+  }
+
+  // Validate that we have a timestamp for when events were last updated
+  const timestampStr = newsNavlink.data("latest-event");
```
Comment (veelkoov) [inline symfony/assets/scripts/entry/general.ts:46]: I understand that Static.ts/DATA_BRIDGE stuff cannot be used here, but I wanted to mention it just to make sure you are aware of it.

File: symfony/src/Repository/EventRepository.php (hunk 1 above)
Comment (veelkoov) [reply symfony/src/Repository/EventRepository.php]: Now I get it. You are doing a `SELECT MAX()` which returns null if there are no rows. I am sorry, I didn't catch that earlier. So now let me think what will happen if a null will get returned....

File: symfony/assets/scripts/class/Storage.ts [hunk 6]
```diff
@@ -36,6 +36,15 @@ export default class Storage {
     this.save(key, value ? "1" : "0", expireSeconds);
   }
 
+  public static remove(key: string): void {
+    try {
+      localStorage.removeItem(this.dataPath(key));
+      localStorage.removeItem(this.expiresPath(key));
+    } catch (e) {
+      console.log("Failed to remove data.", e);
```
Comment (veelkoov) [inline symfony/assets/scripts/class/Storage.ts:44]: Another question from a frontend-inexperienced person. I know that this was moved, but it made me notice that I've used `log` here. Should we switch to `error`/`warn`/`info` instead?

File: symfony/src/Repository/EventRepository.php [hunk 7]
```diff
@@ -51,4 +54,23 @@ public function getRecent(array $types = []): array
...
+                ->enableResultCache(3600)
+                ->getSingleScalarResult();
+            if (null === $resultData) {
+                return null;
+            }
+
+            return UtcClock::at(Enforce::nString($resultData));
+        } catch (NoResultException) {
+            return null;
+        }
```
Comment (veelkoov) [inline symfony/src/Repository/EventRepository.php:74]: So... sorry for my mistake, the exception would not be thrown in this case. You probably can switch to `Enforce::string`, because null won't happen after line 67. I am adding an issue to improve phpdocs/interface of UtcClock, because I have no idea what was I thinking allowing null/false and returning current time in that case.

File: symfony/src/Repository/EventRepository.php (hunk 1 above)
Comment (TangerineFloof) [reply symfony/src/Repository/EventRepository.php]: FWIW after your initial comment, I tested this code after deleting all of the rows from the DB. It returned `null` without throwing an error (which was why I explored the code, since the documentation says that it should throw an error).

File: symfony/assets/scripts/class/Visitor.ts (hunk 2 above)
Comment (TangerineFloof) [reply symfony/assets/scripts/class/Visitor.ts:37]: Either approach is valid! It's just my personal preference to always do `export default` at the end of the file.

The reason I chose to do that is because when you compose the default export in another function, if you do that a couple of times, it starts affecting the indentation. You would see this a _lot_ in legacy class components React:
```typescript
// FILE: MyApp.tsx
export default connectRedux(myFunc)(
  withAuth(
    withCache(
      withFeatureFlags(
        class MyApp extends React.Component {
        }
      )
    )
  )
);
```
That meant that every time you added another function, it would completely reformat the whole file, leading to a lot of Git issues. By comparison, you don't have that same issue when it's `export default` at the end of the file.

I also find it just a little easier to read through the file and figure out what's the default export when it's always in a consistent location, vs. when it's mixed in with the file.

Ultimately though, this is just my personal preference. I don't think there's any real majority convention in JavaScript.

File: symfony/assets/scripts/entry/general.ts (hunk 3 above)
Comment (TangerineFloof) [reply symfony/assets/scripts/entry/general.ts:38]: More than anything, you're definitely the lead for the codebase, so if you'd prefer any particular conventions I am happy to adopt them.

My personal thoughts are that it might not be necessary in TypeScript. In pure JavaScript, I think there's a lot of value — what functions are available are totally different if it's a vanilla DOM node vs. jQuery object. But in TypeScript, the static compiler and the IDE will tell you if it's jQuery or not, so the Hungarian-notation doesn't offer quite the same value. Without a linter to keep it accurate, it could also become inaccurate over time (we refactor `$myDiv` to no longer be a jQuery object but forget to update the name).

File: symfony/assets/scripts/entry/general.ts (hunk 5 above)
Comment (TangerineFloof) [reply symfony/assets/scripts/entry/general.ts:46]: Thank you! I did see that, and I thought at first about doing a similar approach. I went with the `data` attribute approach _mostly_ because it'd been a while since I'd used that technology (and it sounded fun), and then a little bit because it'd keep the data a bit more isolated. It won't be globally available/get in the way, the data is scoped just to this use case.

It was mostly the former reason tho haha

File: symfony/assets/scripts/entry/general.ts (hunk 4 above)
Comment (TangerineFloof) [reply symfony/assets/scripts/entry/general.ts:74]: There were a couple of reasons for my choice here:
1. Performance (you can bypass the string parsing portion)
2. Personal preference for doing one operation at a time where possible
    * SRE principle, but also when something fails/causes an error, you get a more specific location for what went wrong rather than "this string is invalid"
3. Strings don't offer as much type safety/static analysis opportunities
    * TypeScript won't parse/validate the string — so it's easier to have something sneak in
    * A theoretical ESLint plugin could validate that all of your CSS classes must exist in an **.scss** file (look out for typos). This would be pretty trivial to do when you're using functions like `.addClass`, but a lot harder when you're using a composite string (you'd need to teach ESLint how to parse the string as HTML)
4. Easier to refactor/change in the future (at least IMO), since you can change the nodes around without needing to have it all be in a single string

You're absolutely right though — this is happening one time, at page load, and this isn't a complex HTML string. This is just a personal preference, and is also me wanting to ensure I have a consistent style throughout a codebase (vs. a style where I could use `.append("<div />")` for simple things and then `document.createElement` for complex things).

Comment (TangerineFloof) [reply symfony/assets/scripts/entry/general.ts:74]: As for the block-element-in-inline element thing, that's probably not perfect. It's mostly a habit I've picked up over the years. But changing something from block → inline is so easily done. and I don't think(?) it has any impact on a11y.

In this case, the `<div />` is being absolutely positioned anyway so it's sorta outside the primary document block/inline layout anyways.

tl;dr tho, bad habits die hard

File: symfony/assets/scripts/class/Storage.ts (hunk 6 above)
Comment (TangerineFloof) [reply symfony/assets/scripts/class/Storage.ts:44]: It wouldn't hurt, but I don't think it's a huge problem. For smaller projects where your console volume is relatively low anyways, I don't know that it matters as much in JS as it does on the terminal, where `log` goes to stdout and `error` goes to stderr and maybe you're piping them to different locations.

It's probably more worth investing in a replacement/wrapper for `console`. Typically, you'll want to have these kinds of logs on your development builds but then have them noop or be removed by the build process for prod, since otherwise this is all happening for the customer too. That's usually what frontend JS developers more focus on.

File: symfony/assets/scripts/entry/general.ts (hunk 3 above)
Comment (veelkoov) [reply symfony/assets/scripts/entry/general.ts:38]: Fair point. I started creating a document explaining code practices/naming/deprecations and this may become a new entry there.

**Comment by TangerineFloof** [comment]
Closing in favor of #235. That PR contains many of the changes found in this branch. It does not include the unread news badge. Following #235 being merged, I'll create a new branch that brings this code into the new system.
